paddle/phi/kernels/onednn/add_n_kernel.cc

```diff
@@ -0,0 +1,141 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "paddle/phi/kernels/add_n_kernel.h"
+#include "paddle/phi/backends/onednn/onednn_reuse.h"
+#include "paddle/phi/core/kernel_registry.h"
+
+namespace phi {
+namespace funcs {
+template <typename T>
+class AddOneDNNHandler
+    : public phi::funcs::OneDNNHandlerNoCachingT<T, dnnl::sum> {
+ public:
+  AddOneDNNHandler(dnnl::engine engine,
+                   const phi::Place& cpu_place,
+                   const std::vector<const TensorBase*>& x,
+                   DenseTensor* z)
+
+      : phi::funcs::OneDNNHandlerNoCachingT<T, dnnl::sum>(engine, cpu_place),
+        num_inputs_(0) {
+    auto dst_tz = phi::vectorize<int64_t>(z->dims());
+    auto src_tz = dst_tz;
+
+    std::vector<dnnl::memory::desc> srcs_md;
+    srcs_md.reserve(x.size());
+    for (size_t i = 0; i < x.size(); i++) {
+      auto* input_it = (static_cast<const DenseTensor*>(x[i]));
+      if (input_it->numel() == 0) {
+        continue;
+      }
+      srcs_md.push_back(input_it->mem_desc());
+      ++num_inputs_;
+    }
+    std::vector<float> scales(num_inputs_, 1.0f);
+
+    auto dst_md = dnnl::memory::desc(
+        dst_tz, OneDNNGetDataType<T>(), OneDNNMemoryFormat::any);
+
+    this->AcquireForwardPrimitiveDescriptor(dst_md, scales, srcs_md);
+  }
+
+  // (jczaja) sum oneDNN prim is not having .desc attribute so
+  // we cannot use base AcquireForwardPrimitiveDescriptor
+  void AcquireForwardPrimitiveDescriptor(
+      const dnnl::memory::desc& dst_md,
+      const std::vector<float>& scales,
+      const std::vector<dnnl::memory::desc>& srcs_md) {
+    this->fwd_pd_.reset(
+        new dnnl::sum::primitive_desc(dst_md, scales, srcs_md, this->engine_));
+  }
+
+  std::shared_ptr<dnnl::memory> AcquireSrcMemory(const DenseTensor* input,
+                                                 int i) {
+    const T* input_data = input->data<T>();
+    return this->AcquireMemoryFromPrimitive(this->fwd_pd_->src_desc(i),
+                                            to_void_cast<T>(input_data));
+  }
+
+  using phi::funcs::OneDNNHandlerNoCachingT<T, dnnl::sum>::AcquireDstMemory;
+
+  std::shared_ptr<dnnl::memory> AcquireDstMemory(void) {
+    return this->AcquireMemoryFromPrimitive(this->fwd_pd_->dst_desc());
+  }
+
+  inline int GetNumInputs(void) { return num_inputs_; }
+
+ private:
+  int num_inputs_;
+};
+}  // namespace funcs
+
+template <typename T, typename Context>
+void AddNKernel(const Context& dev_ctx,
+                const std::vector<const TensorBase*>& x,
+                DenseTensor* out) {
+  PADDLE_ENFORCE_EQ(
+      dev_ctx.GetPlace().GetType() == phi::AllocationType::CPU,
+      true,
+      phi::errors::PreconditionNotMet("Operator DNNL Sum must use CPUPlace"));
```

Comment:
DNNL is deprecated, please replace with oneDNN.
I'd change `Operator DNNL Sum` either to `oneDNN Sum primitive` or `oneDNN AddN kernel`.